api/v1alpha1/provisioning_validation.go

```diff
@@ -0,0 +1,206 @@
+package v1alpha1
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	log = ctrl.Log.WithName("provisioning_validation")
+)
+
+/*
+   Validate the contents of the provisioning resource
+
+   Managed:
+   "ProvisioningInterface"
+   "ProvisioningIP"
+   "ProvisioningNetworkCIDR"
+   "ProvisioningDHCPRange"
+   "ProvisioningOSDownloadURL"
+
+   Unmanaged:
+   "ProvisioningInterface"
+   "ProvisioningIP"
+   "ProvisioningNetworkCIDR"
+   "ProvisioningOSDownloadURL"
+
+   Disabled:
+   "ProvisioningOSDownloadURL"
+*/
+func (prov *Provisioning) Validate() error {
+	provisioningNetworkMode := prov.getProvisioningNetworkMode()
+	log.V(1).Info("provisioning network", "mode", provisioningNetworkMode)
+
+	var errs []error
+
+	// They all use provisioningOSDownloadURL
+	if err := validateProvisioningOSDownloadURL(prov.Spec.ProvisioningOSDownloadURL); err != nil {
+		errs = append(errs, err...)
+	}
+
+	// Disabled does not need to check network settings.
```

Comment:
I think we still want to validate them in the disabled case *if they're present*. If a user provides a ProvisioningIP, we want to make sure they gave us the ProvisioningNetworkCIDR *and* that the ProvisioningIP is located in that CIDR.

Sorry for the confusion because it's really weird this stuff says Provisioning when the network is disabled. This was a kludge: when there's no network, the user still needs to give us a VIP on a machine network, and they tell us which one by stuffing it in ProvisioningNetworkCIDR.

Reply:
OK will do.  Thanks Stephen!